src/core/compiler.jl

```diff
@@ -288,49 +229,144 @@ function build_model_info(input_expr)
+    vi = model_info[:main_body_names][:vi]
+    ctx = model_info[:main_body_names][:ctx]
+    sampler = model_info[:main_body_names][:sampler]
+    temp_right = gensym(:temp_right)
+    out = gensym(:out)
+    lp = gensym(:lp)
+    preprocessed = gensym(:preprocessed)
+    assert_ex = :(Turing.Core.assert_dist($temp_right, msg = $(wrong_dist_errormsg(@__LINE__))))
+    if left isa Symbol || left isa Expr
+        ex = quote
+            $temp_right = $right
+            $assert_ex
+            $preprocessed = Turing.Core.@preprocess($arg_syms, Turing.getmissing($model), $left)
+            if $preprocessed isa Turing.VarName
+                $out = Turing.Inference.tilde($ctx, $sampler, $temp_right, $preprocessed, $vi)
+                $left = $out[1]
+                $vi.logp += $out[2]
+            else
+                $vi.logp += Turing.Inference.tilde($ctx, $sampler, $temp_right, $preprocessed, $vi)
+            end
+        end
+    else
+        ex = quote
+            $temp_right = $right
+            $assert_ex
+            $vi.logp += Turing.Inference.tilde($ctx, $sampler, $temp_right, $left, $vi)
         end
     end
-    if length(fargs_default_values) == 0
-        tent_arg_defaults_nt = :(NamedTuple())
+    return ex
+end
+
+function dot_tilde(left, right, model_info)
+    arg_syms = Val((model_info[:arg_syms]...,))
+    model = model_info[:main_body_names][:model]
+    vi = model_info[:main_body_names][:vi]
+    ctx = model_info[:main_body_names][:ctx]
+    sampler = model_info[:main_body_names][:sampler]
+    out = gensym(:out)
+    temp_left = gensym(:temp_left)
+    temp_right = gensym(:temp_right)
+    preprocessed = gensym(:preprocessed)
+    lp = gensym(:lp)
+    assert_ex = :(Turing.Core.assert_dist($temp_right, msg = $(wrong_dist_errormsg(@__LINE__))))
+    if left isa Symbol || left isa Expr
+        ex = quote
+            $temp_right = $right
+            $assert_ex
+            $preprocessed = Turing.Core.@preprocess($arg_syms, Turing.getmissing($model), $left)
+            if $preprocessed isa Turing.VarName
+                $temp_left = $left
+                $out = Turing.Inference.dot_tilde($ctx, $sampler, $temp_right, $temp_left, $preprocessed, $vi)
+                $left .= $out[1]
+                $vi.logp += $out[2]
+            else
+                $temp_left = $preprocessed
+                $vi.logp += Turing.Inference.dot_tilde($ctx, $sampler, $temp_right, $temp_left, $vi)
+            end
+        end
     else
-        tent_arg_defaults_nt = :($(fargs_default_values...),)
+        ex = quote
+            $temp_left = $left
+            $temp_right = $right
+            $assert_ex
+            $vi.logp += Turing.Inference.dot_tilde($ctx, $sampler, $temp_right, $temp_left, $vi)
+        end
     end
-    model_info[:args] = fargs
-    model_info[:tent_arg_defaults_nt] = tent_arg_defaults_nt
-    return model_info
+    return ex
 end
 
+is_number_or_array_type(T::Type{<:AbstractFloat}) = true
```

Comment:
Please consider replacing this function with something like
```julia
const FloatOrArray = Union{AbstractFloat, AbstractArray}
```
and `isa FloatOrArray` wherever relevant. It's just a bit cleaner imho.